contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java

```diff
@@ -58,215 +68,156 @@
+        processor[fieldCount++] = new Optional(new FmtDate(format == null ? "dd/MM/yyyy" : format));
       } else {
-        processors[i] = new Optional();
+        processor[fieldCount++] = new Optional();
       }
     }
-
+    return processor;
   }
 
   @Override
   public String convert(Object tuple)
   {
+    incomingTuplesCount++;
+    if (tuple == null) {
+      errorTupleCount++;
+      logger.error(" Null tuple", tuple);
+      return null;
+    }
     try {
       StringWriter stringWriter = new StringWriter();
       ICsvBeanWriter beanWriter = new CsvBeanWriter(stringWriter, preference);
       beanWriter.write(tuple, nameMapping, processors);
       beanWriter.flush();
       beanWriter.close();
+      emittedObjectCount++;
       return stringWriter.toString();
     } catch (SuperCsvException e) {
-      logger.debug("Error while converting tuple {} {}",tuple,e.getMessage());
+      logger.error("Error while converting tuple {} {}", tuple, e.getMessage());
+      errorTupleCount++;
     } catch (IOException e) {
       DTThrowable.rethrow(e);
     }
     return null;
   }
 
-  public static class Field
-  {
-    String name;
-    String format;
-    FIELD_TYPE type;
-
-    public String getName()
-    {
-      return name;
-    }
-
-    public void setName(String name)
-    {
-      this.name = name;
-    }
-
-    public FIELD_TYPE getType()
-    {
-      return type;
-    }
-
-    public void setType(String type)
-    {
-      this.type = FIELD_TYPE.valueOf(type);
-    }
-
-    public String getFormat()
-    {
-      return format;
-    }
-
-    public void setFormat(String format)
-    {
-      this.format = format;
-    }
-  }
-
   /**
-   * Gets the array list of the fields, a field being a POJO containing the name
-   * of the field and type of field.
+   * Get the schema
    * 
-   * @return An array list of Fields.
+   * @return schema
    */
-  public ArrayList<Field> getFields()
+  public String getSchema()
   {
-    return fields;
+    return schema;
   }
 
   /**
-   * Sets the array list of the fields, a field being a POJO containing the name
-   * of the field and type of field.
+   * Set the schema
    * 
-   * @param fields
-   *          An array list of Fields.
+   * @param schema
    */
-  public void setFields(ArrayList<Field> fields)
+  public void setSchema(String schema)
   {
-    this.fields = fields;
+    this.schema = schema;
   }
 
-  /**
-   * Gets the delimiter which separates fields in incoming data.
-   * 
-   * @return fieldDelimiter
-   */
-  public int getFieldDelimiter()
+  @VisibleForTesting
+  public long getErrorTupleCount()
   {
-    return fieldDelimiter;
+    return errorTupleCount;
   }
 
-  /**
-   * Sets the delimiter which separates fields in incoming data.
-   * 
-   * @param fieldDelimiter
-   */
-  public void setFieldDelimiter(int fieldDelimiter)
+  @VisibleForTesting
+  public long getEmittedObjectCount()
```

Comment:
You can make this protected? As the testcase will be in the same package, it'll be accesible via protected as well.